### DISCUSSION TITLE
Provide a trailing slash option

### DISCUSSION THREAD

**Discussion by KyleAMathews** [gatsbyjs/gatsby#31372] [RFC] [answered]
Browsers and bots treat pages with and without trailing slashes as unique pages (e.g. `/about/` vs. `/about`) but Gatsby doesn't.

This makes it easy to accidentally create multiple versions of pages in search engine's crawls of your site.

@jon-sully has a good overview of the problem and possible current approaches to solving it https://jonsully.net/blog/trailing-slashes-and-gatsby/

There are popular Gatsby plugins to both [remove slashes](https://www.gatsbyjs.com/plugins/gatsby-plugin-remove-trailing-slashes/?=slash) and [add them](https://www.gatsbyjs.com/plugins/gatsby-plugin-force-trailing-slashes/?=slash).

The approach these plugins take has the shortcoming that they can only control the paths that are created at build-time. There's no way to force slash or no slash for links created with the `<Link>` component at runtime.

So we should handle this at the framework level and give people the option to force all links to include/disclude slashes and handle this at build & runtime.

For build-time, we would modify the [`createPage` action](https://github.com/gatsbyjs/gatsby/blob/c8db78fad43294413b950c49205adcf3dbb6dd1e/packages/gatsby/src/redux/actions/public.js#L177) to modify the path as necessary if the option is set. For runtime, we would do the same thing in the [Link component](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-link/src/index.js) as well as the `navigate` function.

It's also common to want to enforce link creation in markdown or CMSs but this seems difficult to influence from within Gatsby. 

Previous explorations:
- https://github.com/gatsbyjs/gatsby/pull/30230

**Comment by LekoArts** [comment, marked as answer]
This is now released in Gatsby 4.7: https://www.gatsbyjs.com/docs/reference/release-notes/v4.7/#trailingslash-option
